diff -Nur teacup-1.1/hostsetup.py teacup-1.1-modified/hostsetup.py
--- teacup-1.1/hostsetup.py
+++ teacup-1.1-modified/hostsetup.py
@@ -858,7 +858,7 @@
 
     if htype == 'FreeBSD':
         # record the number of reassembly queue overflows
-        run('sysctl net.inet.tcp.reass.overflows')
+        #run('sysctl net.inet.tcp.reass.overflows')
 
         # disable auto-tuning of receive buffer
         run('sysctl net.inet.tcp.recvbuf_auto=0')
@@ -869,6 +869,8 @@
         # send and receiver buffer max (2MB by default on FreeBSD 9.2 anyway)
         run('sysctl net.inet.tcp.sendbuf_max=2097152')
         run('sysctl net.inet.tcp.recvbuf_max=2097152')
+        run('sysctl net.inet.tcp.recvspace=655360')
+        run('sysctl net.inet.tcp.sendspace=327680')
 
         # clear host cache quickly, otherwise successive TCP connections will
         # start with ssthresh and cwnd from the end of most recent tcp
diff -Nur teacup-1.1/loggers.py teacup-1.1-modified/loggers.py
--- teacup-1.1/loggers.py
+++ teacup-1.1-modified/loggers.py
@@ -41,6 +41,7 @@
 from getfile import getfile
 from runbg import runbg
 
+import random
 
 ## Collect all the arguments (here basically a dummy method because we
 ## don't used the return value)
@@ -505,7 +506,7 @@
                     snap_len, interface, file_name, tcpdump_filter)
         pid = runbg(tcpdump_cmd)
 
-        name = 'tcpdump-' + interface
+        name = 'tcpdump-' + interface + str(random.randint(0, 50000))
         #bgproc.register_proc(env.host_string, name, '0', pid, file_name)
         bgproc.register_proc_later(
             env.host_string,
